stencila/pyla/servers.py

```diff
@@ -0,0 +1,289 @@
+    stream = get_stream_buffer(stream)
+    stream.write(message)
+    stream.flush()
+
+
+def stream_read(stream: StreamType, count: int) -> bytes:
+    """Abstract reading from stream to work with IO (buffered/unbuffered) and sockets."""
+    if isinstance(stream, socket):
+        return stream.recv(count)
+
+    return io_read(stream, count)
+
+
+def stream_write(stream: StreamType, message: bytes) -> None:
+    """Abstract writing to stream to work with IO (buffered/unbuffered) and sockets."""
+    if isinstance(stream, socket):
+        stream.send(message)
+    else:
+        io_write(stream, message)
+
+
+def message_read(stream: StreamType) -> str:
+    """Read a length-prefixed message from the stream"""
+    message_len = read_length_prefix(stream)
+    return stream_read(stream, message_len).decode('utf8')
+
+
+def message_write(stream: StreamType, message: str) -> None:
+    """Write a length-prefixed message to the stream."""
+    bites = message.encode('utf8')
+    stream_write(stream, encode_int(len(bites)))
+    stream_write(stream, bites)
+
+
+class JsonRpcErrorCode(enum.Enum):
+    """
+    Error codes defined in JSON-RPC 2.0
+
+    Codes -32000 to -32099	are reserved for implementation-defined server-errors.
+
+    Python implementation of Executa's
+    [JsonRpcErrorCode](https://github.com/stencila/executa/blob/v1.0.0/src/base/JsonRpcError.ts#L14).
+    """
+
+    """
+    Invalid JSON was received by the server.
+    An error occurred on the server while parsing the JSON text.
+    """
+    ParseError = -32700
+
+    """The JSON sent is not a valid Request object."""
+    InvalidRequest = -32600
+
+    """The method does not exist / is not available."""
+    MethodNotFound = -32601
+
+    """Invalid method parameter(s)."""
+    InvalidParams = -32602
+
+    """Internal JSON-RPC error."""
+    InternalError = -32603
+
+    """Generic server error."""
+    ServerError = -32000
+
+    """Capability error."""
+    CapabilityError = -32005
+
+
+class JsonRpcError(BaseException):
+    """
+    A JSON-RPC error that may be part of a response
+
+    Python implementation of Executa's
+    [JsonRpcError](https://github.com/stencila/executa/blob/v1.0.0/src/base/JsonRpcError.ts).
+    """
+
+    """
+    A number that indicates the error type that occurred.
+    This MUST be an integer.
+    """
+    code: JsonRpcErrorCode
+
+    """
+    A string providing a short description of the error.
+    The message SHOULD be limited to a concise single sentence.
+    """
+    message: str
+
+    """
+    A primitive or structured value that contains additional information about the error.
+    This may be omitted.
+    The value of this member is defined by the server (e.g. detailed error information,
+    nested errors etc.).
+    """
+    data: typing.Any
+
+    def __init__(self, code: JsonRpcErrorCode, message: str, data: typing.Any = None):
+        super().__init__(message)
+        self.code = code
+        self.message = message
```

Comment:
this can be omitted as message is set on the super `__init__` call.

Reply:
Yep, but would need a reordering of the args? - `message` first.

Reply:
That's not necessary.

Reply:
Yep, misread your comment. Fixed now.